irohad/ordering/impl/on_demand_ordering_gate.cpp

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ordering/impl/on_demand_ordering_gate.hpp"
+
+#include "interfaces/iroha_internal/proposal.hpp"
+#include "interfaces/iroha_internal/transaction_batch.hpp"
+
+using namespace iroha;
+using namespace iroha::ordering;
+
+OnDemandOrderingGate::OnDemandOrderingGate(
+    std::shared_ptr<OnDemandOrderingService> ordering_service,
+    std::shared_ptr<transport::OdOsNotification> network_client,
+    rxcpp::observable<BlockRoundEventType> events,
+    std::unique_ptr<shared_model::interface::UnsafeProposalFactory> factory,
+    transport::RoundType initial_round)
+    : ordering_service_(std::move(ordering_service)),
+      network_client_(std::move(network_client)),
+      events_subscription_(events.subscribe([this](auto event) {
+        // exclusive lock
+        std::lock_guard<std::shared_timed_mutex> lock(mutex_);
+
+        visit_in_place(event,
+                       [this](BlockEvent block) {
+                         // block committed, increment block round
+                         current_round_ = {block.height, 1};
+                       },
+                       [this](EmptyEvent empty) {
+                         // no blocks committed, increment reject round
+                         current_round_ = {current_round_.first,
```

Comment:
can we use something else instead of pair. I mean round.first and round.second is hard to understand if you don't know that round is pair of (current round and reject round if I am not mistaken)